### PR TITLE
Allow pandas.Series as properties values

### PR DESCRIPTION
**Pull request by mstabrin**
# Description

I tried to use values from a pandas Series as properties for a points layer, but the ordering got messed up during the creation of the Points.

```
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': [1, 2, 3, 4], 'val': [1, 2, 3, 4], 'grp': [2, 1, 1, 2]})

In [3]: for idx, sdf in df.groupby('grp'):
   ...:     viewer.add_points(sdf[['x', 'y']], size=0.1, properties={'val': sdf['val']})
   ...:     

In [4]: viewer.layers[0].properties
Out[4]: {'val': array([nan,  2.])}

In [5]: viewer.layers[1].properties
Out[5]: {'val': array([ 1., nan])}
```

I expected the correct values to be assigned to the val property:

```
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': [1, 2, 3, 4], 'val': [1, 2, 3, 4], 'grp': [2, 1, 1, 2]})

In [3]: for idx, sdf in df.groupby('grp'):
   ...:     viewer.add_points(sdf[['x', 'y']], size=0.1, properties={'val': sdf['val']})
   ...:     

In [4]: viewer.layers[0].properties
Out[4]: {'val': array([2, 3])}

In [5]: viewer.layers[1].properties
Out[5]: {'val': array([1, 4])} 
```

Therefore, I added the functionality to deal with pandas Series objects in addition to numpy arrays and pandas DataFrames.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [X] the test suite for my feature covers inputs of DataFrames, dictionary of Series, Dictionary of lists, Dictionary of numpy arrays
- [X] all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

Does the documentation need an adjustment for this type of change or is a pandas Series in your eyes equivalent to an array since lists are also not mentioned?
Current documentation:
```
properties (dict {str: array (N,)}, DataFrame) – Properties for each point. Each property should be an array of length N, where N is the number of points.
```

### CODE REVIEW

**Comment by andy-sweet** [comment]
@mstabrin : thanks for the write-up and PR!

First, I'd like to apologize for the confusion between properties/features. The intent is that `properties` and `property_choices` are on the way out and `features` is replacing them. But that's not documented and I'm yet to deprecate `properties` (#3911) - both of those are completely my fault.

I'm a little reluctant about this PR because as indicated by the [features docstring](https://github.com/napari/napari/blob/2ca50d4de1e343766edd56a6e68fd6711956c3a5/napari/layers/points/points.py#L541), we'd like to move towards move of a protocol approach to input values for features/properties rather than explicitly supporting pandas-specific types.

If this is blocking you on napari's current version, there are simple workarounds, which you may already be aware of.

Option 1: index `sdf` so you get a `DataFrame`.

```python
for idx, sdf in df.groupby('grp'):
    viewer.add_points(sdf[['x', 'y']], size=0.1, features=sdf[['val']])     
```

Option 2: convert the series to a numpy array (either with `to_numpy` or `np.array`).

```python
for idx, sdf in df.groupby('grp'):
    viewer.add_points(sdf[['x', 'y']], size=0.1, properties={'val': np.array(sdf['val'])})   
```

I actually think option 2 could be a good alternative implementation for the fix here. That way we support anything that is an array-like (like a list) or anything that supports the array protocol rather than pandas `Series` specifically.

File: napari/layers/utils/layer_utils.py
Comment (andy-sweet) [inline napari/layers/utils/layer_utils.py]:
```suggestion
        features = {key: np.array(value, copy=False) for key, value in features.items()}
```

This should handle all array likes and things that implement `__array__`, which includes `Series`. I think `copy=False` is fine here as we don't make guarantees about copies and this should avoid an unnecessary copy when a `np.ndarray` was passed.

Comment (andy-sweet) [reply napari/layers/utils/layer_utils.py]: Might also be good to include an implementation/explanatory comment like you did with a link to this PR.

File: napari/layers/utils/_tests/test_layer_utils.py
Comment (andy-sweet) [inline napari/layers/utils/_tests/test_layer_utils.py]: I think `pandas.testing.assert_frame_equal` should provide more readable test failures, gives us more options when needed, and is also used elsewhere in the tests.

Comment (andy-sweet) [reply napari/layers/utils/_tests/test_layer_utils.py]: Also, I think `_FeatureTable.from_layer` may be a better test target here than `_validate_features` as it is more likely to survive expected deprecations/changes. No way you could know that, but would be good to change.

File: napari/layers/points/_tests/test_points.py
Comment (andy-sweet) [inline napari/layers/points/_tests/test_points.py]: Optional.

I generally prefer integration tests over unit tests (i.e. this over the similar one in `test_layer_utils.py`), but even though this was the original bug, I think this might be overkill and add mostly unnecessarily test maintenance. Especially if you consider that there should be similar tests for other layer's with properties/features.

As it's relatively simple and already written, I don't mind keeping it for now, so this change is optional.

**Comment by andy-sweet** [comment]
> Does the documentation need an adjustment for this type of change or is a pandas Series in your eyes equivalent to an array since lists are also not mentioned?

There will hopefully be quite a few documentation updates soon, though I've been saying that for a while. Depending on the fix in this PR, you should feel free to update the docstrings for `features` and `properties`, but should also feel free not to, as there are a lot and they may change soon anyway.

**Comment by brisvag** [comment]
> I actually think option 2 could be a good alternative implementation for the fix here. That way we support anything that is an array-like (like a list) or anything that supports the array protocol rather than pandas `Series` specifically.

Definitely agree on this one, so we avoid having to add a million if statements :)

**Comment by mstabrin** [comment]
Thank you for your great input @andy-sweet and @brisvag!
I like the general np.array approach. I did not want to change too much in the beginning so I kept the fix related to my related issue.

- Use np.array to create array_like objects
- I removed the integration test from the `points` layer and the `from_layer` function is now used rather then the `_validate_features` function. Since the `from_layer` function already has a bunch of tests, I only added my failing test case.
- Since properties will be deprecated, I added the test for the properties and for the features keyword.
- I added a comment linking to this pull request

**Review by andy-sweet — APPROVED** [review]
Looks great, I'll merge this in 48 hours unless there are any other comments.

**Comment by mstabrin** [comment]
@andy-sweet Thank you :)

**Review by tlambert03 — APPROVED** [review]

**Comment by andy-sweet** [comment]
As this has another approval, I'm going to merge this now. Thanks again for the fix @mstabrin !

**Pull request merged**
